godtossing/Pos/Binary/GodTossing/Core.hs

```diff
@@ -7,83 +7,105 @@ module Pos.Binary.GodTossing.Core
+                                   <> encode (vcSigningKey vssCert)
+  decode = do
+    enforceSize "VssCertificate" 4
+    key <- decode
+    epo <- decode
+    sig <- decode
+    sky <- decode
+    case recreateVssCertificate key epo sig sky of
+      Left e  -> fail e
+      Right v -> pure v
 
 instance Bi Opening where
-    sizeNPut = labelS "Opening" $ putField getOpening
-    get = label "Opening" $ Opening <$> get
+  encode = encode . getOpening
+  decode = Opening <$> decode
 
 instance Bi GtPayload where
-    sizeNPut = labelS "GtPayload" $ convertToSizeNPut toBi
-      where
-        toBi :: GtPayload -> PokeWithSize ()
-        toBi = \case
-            CommitmentsPayload commMap vssMap ->
-                putWord8S 0 <> putS commMap <> putS (toList vssMap)
-            OpeningsPayload opMap vssMap ->
-                putWord8S 1 <> putS opMap <> putS (toList vssMap)
-            SharesPayload sharesMap vssMap ->
-                putWord8S 2 <> putS sharesMap <> putS (toList vssMap)
-            CertificatesPayload vssMap ->
-                putWord8S 3 <> putS (toList vssMap)
-    get = label "GtPayload" $ do
-        getWord8 >>= \case
-            0 -> liftM2 CommitmentsPayload get getVssCerts
-            1 -> liftM2 OpeningsPayload get getVssCerts
-            2 -> liftM2 SharesPayload get getVssCerts
-            3 -> CertificatesPayload <$> getVssCerts
-            tag -> fail ("get@GtPayload: invalid tag: " ++ show tag)
-          where
-            getVssCerts = HM.fromList . map toCertPair <$> get
-            toCertPair vc = (addressHash $ vcSigningKey vc, vc)
+  encode input = case input of
+    CommitmentsPayload  cmap vss -> encodeListLen 3 <> encode (0 :: Word8)
+                                                    <> encode cmap <> encode (toList vss)
+    OpeningsPayload     omap vss -> encodeListLen 3 <> encode (1 :: Word8)
+                                                    <> encode omap <> encode (toList vss)
+    SharesPayload       smap vss -> encodeListLen 3 <> encode (2 :: Word8)
+                                                    <> encode smap <> encode (toList vss)
+    CertificatesPayload vss      -> encodeListLen 2 <> encode (3 :: Word8)
+                                                    <> encode (toList vss)
+  decode = do
+    len <- decodeListLen
+    tag <- decode @Word8
+    case tag of
+      0 -> do
+        matchSize len "GtPayload.CommitmentsPayload" 3
+        liftM2 CommitmentsPayload decode getVssCerts
+      1 -> do
+        matchSize len "GtPayload.OpeningsPayload" 3
+        liftM2 OpeningsPayload decode getVssCerts
+      2 -> do
+        matchSize len "GtPayload.SharesPayload" 3
+        liftM2 SharesPayload decode getVssCerts
+      3 -> do
+        matchSize len "GtPayload.CertificatesPayload" 2
+        CertificatesPayload <$> getVssCerts
+      _ -> fail ("decode@GtPayload: invalid tag: " <> show tag)
+    where
+      getVssCerts = HM.fromList . map toCertPair <$> decode
```

Comment:
I think you are dead-right; my mistake was copying over the old `Pos.Binary.GodTossing.Core` instances also mirroring their logic fully (which is a good idea generally, but not in this case). It's even better, if possible, to simply use the `HashMap` directly, if we can, as for that one we can control canonical encoding easily.